src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -517,102 +498,158 @@ private unsafe void ReadExactly(byte* ptr, int size)
             return result;
         }
 
-        private void ReceiveSingleDocumentsBatch(DocumentsOperationContext documentsContext, int replicatedItemsCount, int attachmentStreamCount, long lastEtag, IncomingReplicationStatsScope stats)
+        public class DataForReplicationCommand : IDisposable
         {
-            if (_log.IsInfoEnabled)
+            private readonly bool _isReplayTransaction;
+
+            public DataForReplicationCommand(bool isReplayTransaction = false)
             {
-                _log.Info($"Receiving replication batch with {replicatedItemsCount} documents starting with {lastEtag} from {ConnectionInfo}");
+                _isReplayTransaction = isReplayTransaction;
             }
 
-            var incomingReplicationAllocator = new IncomingReplicationAllocator(documentsContext, _database);
-            var sw = Stopwatch.StartNew();
-            Task task = null;
+            internal DocumentDatabase DocumentDatabase { get; set; }
 
-            try
+            internal ConflictManager ConflictManager { get; set; }
+
+            internal string SourceDatabaseId { get; set; }
+
+            internal ArraySegment<ReplicationItem> ReplicatedItems { get; set; }
+
+            internal Dictionary<Slice, ReplicationAttachmentStream> ReplicatedAttachmentStreams { get; set; }
+
+            public TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures { get; set; }
+
+            public Logger Logger { get; set; }
+
+            public void Dispose()
             {
-                using (var networkStats = stats.For(ReplicationOperation.Incoming.Network))
+                if (ReplicatedItems != null)
```

Comment:
ReplicatedItems is a struct, can never be null. You probably want just iterate over it.

Reply:
done